src/shogun/lib/sg_type_traits.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Saatvik Shah
+ */
+
+#include <type_traits>
+
+#ifndef SHOGUN_SG_TYPE_TRAITS_H
+#define SHOGUN_SG_TYPE_TRAITS_H
+/**
+ * A collection of useful type traits
+ */
+
+namespace shogun {
+template <typename T1, typename T2>
+constexpr bool sg_is_same_v = std::is_same<T1, T2>::value;
+// Checks if any one of the types matches
+// Ref: https://stackoverflow.com/a/17032517/3656081
+template<typename T, typename... Rest>
+struct sg_is_any_of : std::false_type {
+};
+
+template<typename T, typename First>
+struct sg_is_any_of<T, First> : std::is_same<T, First> {
+};
+
+template<typename T, typename First, typename... Rest>
+struct sg_is_any_of<T, First, Rest...>
+ : std::integral_constant<bool, std::is_same<T, First>::value ||
+     sg_is_any_of<T, Rest...>::value> {
+};
+
+template<typename T, typename First, typename... Rest>
+constexpr bool sg_is_any_of_v = sg_is_any_of<T, First, Rest...>::value;
+
+template <typename T>
+concept Floating = std::is_floating_point<T>::value;
+template <typename T>
+concept Integral = std::is_integral<T>::value;
+
+// Anything that uses the fill_array* macros
+template<typename T>
+concept RngVectorizable = requires(T* container, index_t len, CRandom* r){
+{ r->fill_array(container, len) };
+} || requires(T* container, index_t len, CRandom* r){
+{ r->fill_array_co(container, len) };
+};
```

Comment:
So this checks if either `r->fill_array(container, len)` or `r->fill_array(container, len)` are valid? Very cool!

Reply:
Yep it really is :)

Reply:
Good stuff, thanks a lot for the work!

For my own understanding, do you have a specific reference to another concept that is defined similarly to this one you have created here (for example, a concept defined in the range-v3 library)?

Reply:
Most of the standard concepts I found in documentation/talks were not so specialized(even in use case) as this one is. 
I'll go through the ranges-v3 repo once - thats a good idea.

Reply:
So `ranges-v3` uses C++ metaprogramming hacks to get a syntax resembling concepts but not exactly so. A few similar concepts as the one I've used here are: [Integral](https://github.com/ericniebler/range-v3/blob/cf8add9f467aac695b9fa38e168b90df36b1a806/include/range/v3/utility/concepts.hpp#L346) which matches the ones I've setup and [Swappable](https://github.com/ericniebler/range-v3/blob/cf8add9f467aac695b9fa38e168b90df36b1a806/include/range/v3/utility/concepts.hpp#L387) which checks that `swap(x, y)` is valid somewhat similar to `RngVectorizable`.

Reply:
All right! Thanks for explaining :-)